internal/engine/experiment/quicping/quicping.go

```diff
@@ -0,0 +1,479 @@
+			sess.Logger().Infof(fmt.Sprintf("response dissection failed: %s", err))
+			out <- responseInfo{raw: resp, t: respTime, err: err}
+			continue
+		}
+		// propagate receive information
+		out <- responseInfo{raw: resp, t: respTime, dstID: hex.EncodeToString(dst), versions: supportedVersions}
+
+		sess.Logger().Infof("PING got response from %s", addr)
+	}
+}
+
+// Run implements ExperimentMeasurer.Run.
+func (m *Measurer) Run(
+	ctx context.Context,
+	sess model.ExperimentSession,
+	measurement *model.Measurement,
+	callbacks model.ExperimentCallbacks,
+) error {
+	host := string(measurement.Input)
+	// allow URL input
+	if u, err := url.ParseRequestURI(host); err == nil {
+		host = u.Host
+	}
+	service := net.JoinHostPort(host, m.config.port())
+	udpAddr, err := net.ResolveUDPAddr("udp4", service)
+	if err != nil {
+		return err
+	}
+	rep := m.config.repetitions()
+	tk := &TestKeys{
+		Domain:      host,
+		Repetitions: rep,
+	}
+	measurement.TestKeys = tk
+
+	// create UDP socket
+	pconn, err := m.config.networkLibrary().ListenUDP("udp", &net.UDPAddr{})
+	if err != nil {
+		return err
+	}
+	defer pconn.Close()
+
+	// set context and read timeouts
+	deadline := time.Duration(rep*2) * time.Second
+	pconn.SetDeadline(time.Now().Add(deadline))
+	ctx, cancel := context.WithTimeout(ctx, deadline)
+	defer cancel()
+
+	sendInfoChan := make(chan requestInfo)
+	recvInfoChan := make(chan responseInfo)
+	pingMap := make(map[string]*pingInfo)
+
+	// start sender and receiver goroutines
+	go m.sender(ctx, pconn, udpAddr, sendInfoChan, sess)
+	go m.receiver(ctx, pconn, recvInfoChan, sess)
+L:
+	for {
+		select {
+		case req := <-sendInfoChan: // a new ping was sent
+			if req.err != nil {
+				tk.Pings = append(tk.Pings, &SinglePing{
+					ConnIdDst:   req.dstID,
+					ConnIdSrc:   req.srcID,
+					Failure:     archival.NewFailure(req.err),
+					Request:     &model.ArchivalMaybeBinaryData{Value: string(req.raw)},
+					RequestTime: formatTime(req.t),
+				})
+				continue
+			}
+			pingMap[req.srcID] = &pingInfo{request: &req}
+
+		case resp := <-recvInfoChan: // a new response has been received
+			if resp.err != nil {
+				// resp failure means we cannot assign the response to a request
+				tk.UnexpectedResponses = append(tk.UnexpectedResponses, makeResponse(&resp))
+				continue
+			}
+			var (
+				ping *pingInfo
+				ok   bool
+			)
+			// match response to request
+			if ping, ok = pingMap[resp.dstID]; !ok {
+				// version negotiation response with an unknown destination ID
+				tk.UnexpectedResponses = append(tk.UnexpectedResponses, makeResponse(&resp))
+				continue
+			}
+			ping.responses = append(ping.responses, &resp)
+
+		case <-ctx.Done():
+			break L
+		}
+	}
+	// transform ping requests into TestKeys.Pings
+	timeoutErr := errors.New("i/o timeout")
+	for _, ping := range pingMap {
+		if ping.request == nil { // this should not happen
+			return errors.New("internal error: ping.request is nil")
+		}
+		if len(ping.responses) == 0 {
```

Comment:
```suggestion
		if len(ping.responses) <= 0 {
```

I know `len` cannot return a negative value by design but I'd still rather to write robust code. Since `len` in principle could return an `int`, I'd like to see `<=` being used here.